### PR TITLE
CallerCallee add jump to source location in disassembly

### PR DESCRIPTION
**Pull request by lievenhey**
this patch adds a context menu that allows the user to select if he wants to open the source code or opens the disassembly and focus the selected location

closes: #515

### CODE REVIEW

**Comment by GitMensch** [comment]
Thank you for working on this.

I did check the location pane in caller/callee tab and do see both context menu entries - but
* in one case (sourcefile deleted) there is no context menu at all, no message, no output in stderr/logging
* in all others the Disassembly entry did not do anything (also no output in stderr/logging), while "open in file" does work and positions to the right place and Disassembly on the function level (all other panes and views) works fine

Checking F6 yields in the "go to line" being shown, placing "115" in there and pressing Enter does not seem to work either.

Note: all of CTRL+F (in both panes) and F6 (only on the source) has a remarkable time (1-2 seconds) to be shown ?!?

In any case: CTRL+G is already taken, what about ALT+G or CTRL+SHIFT+G?

File: src/resultscallercalleepage.cpp [hunk 1]
```diff
@@ -234,21 +233,17 @@ void ResultsCallerCalleePage::onSourceMapContextMenu(QPoint point)
...
-        emit navigateToCode(location.path, location.lineNumber, 0);
-    }
-}
+    connect(viewCallerCallee, &QAction::triggered, this,
+            [this, location] { emit navigateToCode(location.path, location.lineNumber, 0); });
+
+    auto line = index.data(SourceMapModel::FileLineRole).value<Data::FileLine>();
+    auto disassemblyAction = contextMenu.addAction(tr("Disassembly"));
+    connect(disassemblyAction, &QAction::triggered, this, [this, line] {
+        if (m_symbol.isValid())
```
Comment (milianw) [inline src/resultscallercalleepage.cpp:242]: use the validity to enable/disable the action in the first place - to prevent the slot from being called in situations where your code would silently not do anything

File: src/resultscallercalleepage.cpp [hunk 2]
```diff
@@ -234,21 +233,17 @@ void ResultsCallerCalleePage::onSourceMapContextMenu(QPoint point)
...
+            emit jumpToSourceCode(m_symbol, line);
+    });
 
-void ResultsCallerCalleePage::onSourceMapActivated(const QModelIndex& index)
-{
-    const auto location = toSourceMapLocation(index);
-    if (location) {
-        emit navigateToCode(location.path, location.lineNumber, 0);
-    } else {
-        const auto fileLine = index.data(SourceMapModel::FileLineRole).value<Data::FileLine>();
```
Comment (milianw) [inline src/resultscallercalleepage.cpp:249]: this old validation code and message got lost in translation now no?

File: src/resultsdisassemblypage.cpp [hunk 3]
```diff
@@ -253,6 +253,16 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
     ui->sourceCodeView->setMouseTracking(true);
     setCostHeader(ui->sourceCodeView);
 
+    ui->gotoWidget->hide();
+
+    connect(ui->gotoLineEdit, &QLineEdit::returnPressed, this, [this]() {
+        int lineNumber = ui->gotoLineEdit->text().toInt();
+        int offset = m_sourceCodeModel->index(0, SourceCodeModel::SourceCodeLineNumber).data().value<int>();
+        // TODO: range validation, maybe, qt should handle out of range, at least test
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:261]: please verify and then remove this todo

File: src/resultsdisassemblypage.cpp [hunk 4]
```diff
@@ -275,8 +285,24 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
...
+    layout->addWidget(label);
+    auto edit = new QLineEdit(gotoMenuWidget);
+    layout->addWidget(edit);
+    auto gotoAction = new QWidgetAction(this);
+    gotoAction->setDefaultWidget(gotoMenuWidget);
+
+    connect(edit, &QLineEdit::returnPressed, this, [this, edit] {
+        int lineNumber = edit->text().toInt();
+        int offset = m_sourceCodeModel->index(0, SourceCodeModel::SourceCodeLineNumber).data().value<int>();
+        // TODO: range validation, maybe, qt should handle out of range, at least test
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:300]: dito

File: src/resultscallercalleepage.cpp [hunk 5]
```diff
@@ -93,9 +93,9 @@ ResultsCallerCalleePage::ResultsCallerCalleePage(FilterAndZoomStack* filterStack
     ResultsUtil::setupHeaderView(ui->callerCalleeTableView, contextMenu);
     ResultsUtil::setupCostDelegate(m_callerCalleeCostModel, ui->callerCalleeTableView);
     connect(ui->callerCalleeTableView, &QTreeView::activated, this, [this](const QModelIndex& index) {
-        auto symbol = index.data(CallerCalleeModel::SymbolRole).value<Data::Symbol>();
-        if (symbol.isValid())
-            emit jumpToDisassembly(symbol);
+        m_symbol = index.data(CallerCalleeModel::SymbolRole).value<Data::Symbol>();
```
Comment (milianw) [inline src/resultscallercalleepage.cpp:96]: `activated` gets emitted on double click on my system - it's imo meaningless to set some member variable then. instead, the context menu should  check the current index and ask it for its symbol

File: src/resultscallercalleepage.cpp (hunk 1 above)
Comment (milianw) [reply src/resultscallercalleepage.cpp:242]: also, it's imo not nice to have an action called "disassembly" silently not do anything if the disassembly isn't showing anything relating to the symbol at hand. that would happen with this code, or? since it only scrolls - it doesn't reload the data in the disassembly view?

File: src/resultsdisassemblypage.cpp [hunk 6]
```diff
@@ -253,6 +253,16 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
     ui->sourceCodeView->setMouseTracking(true);
     setCostHeader(ui->sourceCodeView);
 
+    ui->gotoWidget->hide();
+
+    connect(ui->gotoLineEdit, &QLineEdit::returnPressed, this, [this]() {
+        int lineNumber = ui->gotoLineEdit->text().toInt();
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:259]: here and a couple times below: const auto

File: src/resultsdisassemblypage.cpp [hunk 7]
```diff
@@ -275,8 +285,24 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
...
+    auto gotoMenuWidget = new QWidget(this);
+    auto layout = new QHBoxLayout(gotoMenuWidget);
+    auto label = new QLabel(QStringLiteral("Goto: "), gotoMenuWidget);
+    layout->addWidget(label);
+    auto edit = new QLineEdit(gotoMenuWidget);
+    layout->addWidget(edit);
+    auto gotoAction = new QWidgetAction(this);
+    gotoAction->setDefaultWidget(gotoMenuWidget);
+
+    connect(edit, &QLineEdit::returnPressed, this, [this, edit] {
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:297]: why only on return pressed? why not on textEdited / textChanged?

File: src/resultsdisassemblypage.cpp [hunk 8]
```diff
@@ -286,6 +312,10 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
         auto* openEditorAction = contextMenu.addAction(QCoreApplication::translate("Util", "Open in Editor"));
         QObject::connect(openEditorAction, &QAction::triggered, &contextMenu,
                          [this, fileLine]() { emit navigateToCode(fileLine.file, fileLine.line, -1); });
+
+        // TODO: fix focus, pressing enter will trigger both actions
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:316]: yes please do that then

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/resultscallercalleepage.cpp (hunk 1 above)
Comment (milianw) [reply src/resultscallercalleepage.cpp:242]: I think I (hopefully?) misunderstand the code. can you clarify: if I select this on a different entry than is currently shown in the disassembly view, it will select that one first and then scroll to the line afterwards?

Comment (lievenhey) [reply src/resultscallercalleepage.cpp:242]: ugh, there is a lot of bad naming going on `ResultsCallerCalleePage::jumpToSourceCode` calls `ResultsPage::onJumpToSourceCode` which calls `ResultsDisassemblyPage::onJumpToDisassembly` which will actually change the file. Afterwords `ResultsPage::onJumpToSourceCode` calls `ResultsDisassemblyPage::jumpToSourceLine`

Comment (lievenhey) [reply src/resultscallercalleepage.cpp:242]: so yeah it will change the file and then scroll to the line

File: src/resultscallercalleepage.cpp (hunk 2 above)
Comment (lievenhey) [reply src/resultscallercalleepage.cpp:249]: yes, fixed it

**Comment by lievenhey** [comment]
I added it also for the disassembly, but it doesn't work there (currently)

File: src/resultscallercalleepage.cpp (hunk 1 above)
Comment (milianw) [reply src/resultscallercalleepage.cpp]: would be nice if you could clean this up eventually

File: src/resultscallercalleepage.h [hunk 9]
```diff
@@ -10,6 +10,7 @@
 
 #include <QWidget>
 
+#include "data.h"
```
Comment (milianw) [inline src/resultscallercalleepage.h:13]: forward declare instead

File: src/resultsdisassemblypage.cpp [hunk 10]
```diff
@@ -274,19 +274,58 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
...
-    connect(ui->sourceCodeView, &QTreeView::customContextMenuRequested, this, [this](QPoint point) {
-        const auto index = ui->sourceCodeView->indexAt(point);
-        const auto fileLine = index.data(SourceCodeModel::FileLineRole).value<Data::FileLine>();
-        if (!fileLine.isValid())
-            return;
+    auto createContextMenu = [](QTreeView* view, auto* model, int role, auto&& addEntries) {
+        auto gotoMenuWidget = new QWidget(view);
+        auto layout = new QHBoxLayout(gotoMenuWidget);
+        layout->setContentsMargins(0, 0, 0, 0);
+        auto label = new QLabel(QCoreApplication::translate("Util", "Goto: "), gotoMenuWidget);
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:281]: you can just use `tr` here, no?

File: src/resultsdisassemblypage.cpp [hunk 11]
```diff
@@ -274,19 +274,58 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
...
+        layout->addWidget(label);
+        auto edit = new QLineEdit(gotoMenuWidget);
+        layout->addWidget(edit);
+        auto gotoAction = new QWidgetAction(view);
+        gotoAction->setDefaultWidget(gotoMenuWidget);
+
+        auto gotoLine = [view, model, role, edit] {
+            const auto lineNumber = edit->text().toInt();
+
+            // model->index()->data is somehow not possible
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:291]: what does this mean?

File: src/resultsdisassemblypage.cpp [hunk 12]
```diff
@@ -274,19 +274,58 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
...
+            const QModelIndex index = model->index(0, role);
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:292]: auto

File: src/resultsdisassemblypage.cpp [hunk 13]
```diff
@@ -274,19 +274,58 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(CostContextMenu* costContextMenu,
...
+            contextMenu.exec(QCursor::pos());
+        });
+    };
+
+    createContextMenu(ui->sourceCodeView, m_sourceCodeModel, SourceCodeModel::SourceCodeLineNumber,
+                      [this](QMenu* menu, const QModelIndex& index) {
+                          const auto fileLine = index.data(SourceCodeModel::FileLineRole).value<Data::FileLine>();
+                          if (fileLine.isValid()) {
+                              auto* openEditorAction =
+                                  menu->addAction(QCoreApplication::translate("Util", "Open in Editor"));
```
Comment (milianw) [inline src/resultsdisassemblypage.cpp:320]: just `tr`?

File: src/resultsdisassemblypage.ui [hunk 14]
```diff
@@ -87,7 +87,7 @@
         </property>
         <property name="icon">
          <iconset theme="go-previous">
-          <normaloff>.</normaloff>.</iconset>
+          <normaloff>../../../.designer/backup</normaloff>../../../.designer/backup</iconset>
```
Comment (milianw) [inline src/resultsdisassemblypage.ui:90]: unrelated changes?

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/resultsdisassemblypage.cpp (hunk 11 above)
Comment (lievenhey) [reply src/resultsdisassemblypage.cpp:291]: If I try `model->index(..)->data()` the compiler throws an error

Comment (lievenhey) [reply src/resultsdisassemblypage.cpp:291]: same if I replace `QModelIndex` with auto.
Error message:
```
hotspot/src/resultsdisassemblypage.cpp:293:46: error: use 'template' keyword to treat 'value' as a
 dependent template name
            const auto offset = index.data().value<int>();
                                             ^
                                             template 
1 error generated.
```

File: src/resultsdisassemblypage.ui (hunk 14 above)
Comment (lievenhey) [reply src/resultsdisassemblypage.ui:90]: huh? Yes no idea what QDesigner is doing

Comment (lievenhey) [reply src/resultsdisassemblypage.ui:90]: maybe it is because I pressed restore changes after it crashed
